### DISCUSSION TITLE
Buffer Exhaustion at High Load

### DISCUSSION THREAD

**Discussion by mmb-davidsmith** [openthread/openthread#9086] [Q&A]
As part of our testing of our border router stability, we've been trying to make sure we can handle high loads well. To do this, one of our tests uses the topology below and then sends a ping with 65k packets from the PC to the nRF device. We adjust the interval between ping messages to increase the load.

```
 ┌───────────────┐         ┌─────────────────┐      ┌──────────────────────────┐
 │               │         │                 │      │                          │
 │  PC (on LAN)  │◄───────►│  OpenThread BR  │◄────►│  OpenThread FTD (nRF52)  │
 │               │    ▲    │                 │   ▲  │                          │
 └───────────────┘    │    └─────────────────┘   │  └──────────────────────────┘
                      │                          │
                      │                          │
                      │                          │
                  Ethernet                     Thread
```

In this setup, we start to see packet loss when we test at a 10ms interval. This is expected because it exceeds what the thread network can realistically handle. We've done some AB testing with the default OpenThread implementation (we're based on an older Gecko stack) vs an implementation with a leaky bucket on the LAN -> Thread path only.

In the default OpenThread implementation we see that we get somewhere in the order of 18% packet loss. In the logs, we see complaints like the following:

```
E(2049583) OPENTHREAD:[C] Platform------: Failed to copy to OpenThread message: NoBufs
E(2049593) OPENTHREAD:[C] Platform------: Failed to allocate OpenThread message
I(2051073) OPENTHREAD:[N] MeshForwarder-: Dropping rx lowpan HC frame, error:NoBufs, len:44, src:0x5000, dst:0x0c00, sec:yes
```

When we introduce our leaky bucket implementation which sizes the bucket at 90% of the total buffers available in OpenThread, we see that this packet loss drops to a value around 4-5%. We also don't see any of the log messages above, but we do see that we're actively dropping packets.

Currently we're running with `OPENTHREAD_CONFIG_NUM_MESSAGE_BUFFERS` set to `256` to mimic what the posix border router configuration uses. We were using a higher number before, but that seemed to cause some instability and just hides a network issue the upstream should recognize earlier rather than later.

My working theory is that because the LAN side is such a faster interface, we're saturating the message buffer pool with outgoing messages once we go past what the thread network can handle. Because the message buffer pool is shared, we're starving the Thread -> LAN path of buffers to process the responses. Adding this bucket which ensures the LAN -> Thread path drops packets if the bucket is close to full then allows room for these responses and other network maintenance messages to happen.

Do these observations and hypothesis make sense? If so, is it worth adding something into the core implementation to allow for some better balancing of the buffer allocations in heavy traffic situations.

**Comment by jwhui** [comment]
@mmb-davidsmith , thanks for sharing!

My initial thought is that this all makes sense. As you stated, because the LAN -> Thread path is high bandwidth, it is possible to starve incoming traffic from other lower bandwidth sources. I think we should implement some mechanism to enforce fairness between the different traffic sources.

Would you be willing to share more details about your leaky bucket implementation? If it's easier, feel free to share the source code itself.

**Reply by mmb-davidsmith** [reply]
@jwhui - It's a very basic leaky bucket. We're using LWIP for the rest of the networking, so it wraps OpenThread in a netif. In the netif, there's a transmit function and near the top of that transmit function, I have the following code.

```
    otMessageGetBufferInfo(esp_openthread_get_instance(), &buffer_info);
    if (buffer_info.mFreeBuffers < buffer_info.mTotalBuffers / 10) {
        // Make sure we maintain at least 10% of buffers free for other things (eg. RX, network maintenance, etc.)
        //TODO: Log this dropped frame to a counter somewhere, possibly put into event log 
        return ESP_ERR_NO_MEM;
    }
```

It's not the most efficient right now given the `otMessageGetBufferInfo` gets a lot more info than the required `mTotalBuffers` and `mFreeBuffers`, but if OpenThread were to implement something for this it could implement more efficient accessors for this info or just call the two lines that populate this directly.

In terms of the 10%, we haven't really figured out what the right number is. The 10% was an easy first pass that showed a clear improvement. This code will also fail if the total buffers is too small, I just happened to know in my initial testing that the value would be > 10.

**Reply by mmb-davidsmith** [reply]
I should also note that ultimately this blocks a call to `otIp6Send`.

**Reply by mmb-davidsmith** [reply]
Testing in the reverse direction `Thread -> BR -> LAN` results in a lot of dropped packets on the LAN side as it processes responses, so a final solution would need to be a bit more intelligent in giving some sort of fair or guaranteed minimum access to the message buffer pool.

**Comment by abtink** [comment]
@mmb-davidsmith , do you enable and use `OPENTHREAD_CONFIG_MAX_FRAMES_IN_DIRECT_TX_QUEUE` ?

The `OPENTHREAD_CONFIG_MAX_FRAMES_IN_DIRECT_TX_QUEUE` would limit number of frames in tx queue which effectively limits number of allocated messages that can be added to tx queue for direct tx.

For reference this is PR adding this: 
- https://github.com/openthread/openthread/pull/7908

**Reply by jwhui** [reply]
I'm not sure `OPENTHREAD_CONFIG_MAX_FRAMES_IN_DIRECT_TX_QUEUE` addresses fairness. If packets are coming in from a given source at a high rate, that source is like to occupy any available message buffer first.

**Reply by abtink** [reply]
Correct. This only limits the number of outbound messages in OT tx queue, i.e. drops a new direct tx request and frees the `Message` if there are already too many messages in the TX queue.  It won't impact if other components are allocating/exhausting message pool (specially if the app/next layer allocates and uses `otMessage` as well).

In OT core, we limit the the number of messages in TX queue in 3 different ways:
- `OPENTHREAD_CONFIG_MAX_FRAMES_IN_DIRECT_TX_QUEUE` mechanism (limits number of direct tx frames to 100).
- `OPENTHREAD_CONFIG_DELAY_AWARE_QUEUE_MANAGEMENT_ENABLE` which uses "time-in-queue" to drop aged msgs.
- [`MeshForwarder::EvictMessage()`](https://github.com/openthread/openthread/blob/b82cfddd683f065dc297ce04930045314e7053a8/src/core/thread/mesh_forwarder_ftd.cpp#L203-L265) which 
   - First tries to drop messages with lower proirity (if it can find one).
   - Then evicts the oldest indirect message queued for an sleepy child (of same or higher priority).

**Reply by mmb-davidsmith** [reply]
I haven't tried these things yet. I may be on a version that's too old to have these features as I don't see any references show up in a quick grep. As has been discussed, this doesn't ensure any sort of fairness though. 

Most of my testing has been in the LAN -> Thread direction. I've just started the Thread -> LAN testing and see the rate limiting is triggered there when using an interval of 10ms. I haven't been able to dig into it too deeply yet, but this surprised me as I would have thought the 802.15.4 data rate would limit how quickly packets could be received. I'm investigating to see if there's any optimization needed on that path and may update if there's something going on in the Thread stack for this (ie, rate limiting is needed or we service the serial port much more quickly than we forward packets to the upper IP stack)

**Reply by abtink** [reply]
I expect these to help in "LAN -> Thread" direction (i.e., if we forward a lot of traffic from infra to Thread to be sent over mesh, they limit number of message in tx (6lowpan) queue).

They latest one of these feature is https://github.com/openthread/openthread/pull/7908 which was merged Jul 27 of last year. I think the other ones are earlier than this one.

**Reply by jwhui** [reply]
I'm not yet sure that these mechanisms address @mmb-davidsmith 's concern.

- `OPENTHREAD_CONFIG_MAX_FRAMES_IN_DIRECT_TX_QUEUE` is applied to the send queue without considering traffic source.
- `OPENTHREAD_CONFIG_DELAY_AWARE_QUEUE_MANAGEMENT_ENABLE` will help in cases where the traffic sender implements congestion control mechanisms (e.g. TCP).
- `EvictMessage()` only works if the response has higher priority than request.

Thoughts?

**Reply by mmb-davidsmith** [reply]
I think that the `OPENTHREAD_CONFIG_MAX_FRAMES_IN_DIRECT_TX_QUEUE` might be able to come close to what I'm doing right now but does have the difference that it's going to impact internal OpenThread traffic. Right now I'm limiting only the traffic being routed into the network from an external source.

Ultimately I think we would want to set up some sort of system for both the RX and TX paths to make sure they can't consume 100% of the buffers so they don't start to negatively impact each other.

**Reply by jwhui** [reply]
> Right now I'm limiting only the traffic being routed into the network from an external source.

I thought this was the key part of your original thought, to ensure fairness of servicing traffic received from the different interfaces (e.g. LAN vs. Thread).

> Ultimately I think we would want to set up some sort of system for both the RX and TX paths to make sure they can't consume 100% of the buffers so they don't start to negatively impact each other.

Right now, we have priority based mechanisms to ensure things like MLE traffic are not starved out. We'll probably want to look into something like FQ-CoDel [RFC 8290](https://datatracker.ietf.org/doc/html/rfc8290), but we'll have to make sure it can fit within the resource constraints we expect for Thread devices.

**Reply by mmb-davidsmith** [reply]
Jonathan, you're correct. The key thing is I want to limit the traffic from specific interfaces to be fair. My comment was trying point out that I could probably achieve something *close* with the mechanisms Abtin pointed to, but that there is a difference.

**Comment by mmb-davidsmith** [comment]
Adding some further tracing on the Thread -> LAN side, I see the following sort of behaviour sending 1000 ping messages of length 8 with a 10ms interval from a thread host.

```
dropped frame (total: 256, free: 24, send queue: 0, sixlow: 116)
0 messages waiting inbound, counter: 983
dropped frame (total: 256, free: 24, send queue: 0, sixlow: 116)
0 messages waiting inbound, counter: 984
0 messages waiting inbound, counter: 985
0 messages waiting inbound, counter: 986
0 messages waiting inbound, counter: 987
0 messages waiting inbound, counter: 988
0 messages waiting inbound, counter: 989
0 messages waiting inbound, counter: 990
0 messages waiting inbound, counter: 991
0 messages waiting inbound, counter: 992
dropped frame (total: 256, free: 24, send queue: 0, sixlow: 116)
0 messages waiting inbound, counter: 993
dropped frame (total: 256, free: 24, send queue: 0, sixlow: 116)
0 messages waiting inbound, counter: 994
0 messages waiting inbound, counter: 995
dropped frame (total: 256, free: 24, send queue: 0, sixlow: 115)
0 messages waiting inbound, counter: 996
dropped frame (total: 256, free: 24, send queue: 0, sixlow: 116)
0 messages waiting inbound, counter: 997
0 messages waiting inbound, counter: 998
dropped frame (total: 256, free: 24, send queue: 0, sixlow: 115)
0 messages waiting inbound, counter: 999
dropped frame (total: 256, free: 24, send queue: 0, sixlow: 116)
0 messages waiting inbound, counter: 1000
dropped frame (total: 256, free: 24, send queue: 0, sixlow: 116)
```

the print for the dropped frame is as follows:
```
printf("dropped frame (total: %"PRIu16", free: %"PRIu16", send queue: %"PRIu16", sixlow: %"PRIu16")\n", 
    buffer_info.mTotalBuffers, buffer_info.mFreeBuffers, 
    buffer_info.mIp6Queue.mNumMessages, buffer_info.m6loSendQueue.mNumMessages);
```

The inbound counter indicates that we see all 1000 messages which are sent by the the initiating node. This makes me think that the BR seems to be having issues getting responses out. I would have expected to see some sort of loss on both sides in this case. Does this suggest some other path that I should be investigating?
